liquid-compiler/src/filter.rs

```diff
@@ -1,101 +1,254 @@
-use std::fmt::{self, Debug};
+use std::fmt::{Debug, Display};
 
-use liquid_error;
+use liquid_error::Result;
+use liquid_interpreter::{Context, Expression};
 use liquid_value::Value;
 
-/// Expected return type of a `Filter`.
-pub type FilterResult = Result<Value, liquid_error::Error>;
-
-/// A trait for creating custom tags. This is a simple type alias for a function.
+/// A structure that holds the information of a single parameter in a filter.
+/// This includes its name, description and whether it is optional or required.
 ///
-/// This function will be called whenever the parser encounters a tag and returns
-/// a new [Renderable](trait.Renderable.html) based on its parameters. The received parameters
-/// specify the name of the tag, the argument [Tokens](lexer/enum.Token.html) passed to
-/// the tag and the global [`Language`](struct.Language.html).
-pub trait FilterValue: Send + Sync + FilterValueClone + Debug {
-    /// Filter `input` based on `arguments`.
-    fn filter(&self, input: &Value, arguments: &[Value]) -> FilterResult;
-}
-
-/// Support cloning of `Box<FilterValue>`.
-pub trait FilterValueClone {
-    /// Cloning of `dyn FilterValue`.
-    fn clone_box(&self) -> Box<FilterValue>;
+/// This is the return type in some `FilterReflection` functions.
+pub struct ParameterReflection {
+    pub name: &'static str,
+    pub description: &'static str,
+    pub is_optional: bool,
```

Comment:
Adding or changing fields to this struct is a breaking change.  Do you remember what we decided about that?

Reply:
What we decided was for the `FilterReflection` trait itself, which was to keep it as a trait instead of a struct for it to be extensible.

For the `ParameterReflection`, we didn't actually discuss about what to do. 

Some possibilities:
  - Add a private `var = ()` field and implement a `new` function to initialize the struct. We could add new methods to change eventual new members without a breaking change.
  - Implement new fields in the original `FilterReflection` trait. This would probably become messy, as it would be harder to connect fields to their parameter. This would actually have be the only way to extend the 4 `parameters` struct in the reflection from the #301 draft.
  - Turn this struct into a trait. This would imply creating a new dull structure for every parameter inside each filter, which maybe could be acceptable because we have macros.

Reply:
You're right, we can always rev this via filer reflection